providers/hubspot/internal/custom/concurrent.go

```diff
@@ -0,0 +1,63 @@
+package custom
+
+import (
+	"context"
+	"sync"
+)
+
+// concurrentProcessing executes the provided process function in parallel for each element in the list.
+//
+// All calls share the same context. If any invocation returns a non-nil error,
+// the context is cancelled and the function returns immediately with that error.
+//
+// The process function receives a context, the element, and a mutex to safely
+// synchronize access to shared resources (e.g., maps) across goroutines.
+func concurrentProcessing[T any, P func(context.Context, T, *sync.Mutex) error](
```

Comment:
There's a better version of this in the server repo:

https://github.com/amp-labs/server/blob/main/shared/utils/simultaneously.go

It's better because it lets you specify an upper limit for concurrency. We have to be very careful about exploding goroutines - remember that we process requests in parallel, and that even our backend workers tend to have about 1000 threads which can all potentially run at the same time. So the multiplicity is a real problem for us.

You don't need to copy it verbatim - adapt it to your needs and copy-paste whatever makes sense. But the max parallelism is really important.

Reply:
Also, I worry about the increased possibility here of being rate-limited.